volatility3/framework/symbols/linux/extensions/__init__.py

```diff
@@ -2979,3 +3131,139 @@ def get_content(
+                search key and an element. It's works like the kernel's latch_tree_ops::comp
+                i.e.: comp_function(key, latch_tree_node)
+
+        Returns:
+            latch_tree_node: A pointer to the node matching key or None.
+        """
+        # latch_tree_root >= 4.2 ade3f510f93a5613b672febe88eff8ea7f1c63b7
+
+        # Use the lowest sequence bit as an index for picking which data copy to read
+        if self.seq.has_member("seqcount"):
+            # kernels >= 5.10 0c9794c8b6781eb7dad8e19b78c5d4557790597a
+            sequence = self.seq.seqcount.sequence
+        elif self.seq.has_member("sequence"):
+            # 4.2 <= kernel < 5.10
+            sequence = self.seq.sequence
+        else:
+            raise AttributeError("Unsupported sequence type implementation")
+
+        idx = sequence & 1
+
+        rb_node_ptr = self.tree[idx].rb_node
+        while rb_node_ptr and rb_node_ptr.is_readable():
+            rb_node = rb_node_ptr.dereference()
+            lt_node = self._get_lt_node_from_rb_node(rb_node, idx)
+            c = comp_function(key, lt_node)
+            if c < 0:
+                rb_node_ptr = rb_node.rb_left
+            elif c > 0:
+                rb_node_ptr = rb_node.rb_right
+            else:
+                return lt_node
+
+        return None
+
+
+class kernel_symbol(objects.StructType):
+
+    def _offset_to_ptr(self, off) -> int:
+        layer = self._context.layers[self.vol.layer_name]
+        long_mask = (1 << layer.bits_per_register) - 1
+        return (self.vol.offset + off) & long_mask
+
+    @property
+    def name(self) -> str:
+        if self.has_member("name_offset"):
+            # kernel >= 4.19 and CONFIG_HAVE_ARCH_PREL32_RELOCATIONS=y
+            # See 7290d58095712a89f845e1bca05334796dd49ed2
+            name_offset = self._offset_to_ptr(self.name_offset)
+        elif self.has_member("name"):
+            # kernel < 4.19 or CONFIG_HAVE_ARCH_PREL32_RELOCATIONS=n
+            name_offset = self.member("name")
+        else:
+            raise AttributeError("Unsupported kernel_symbol type implementation")
+
+        layer = self._context.layers[self.vol.layer_name]
+        name_bytes = layer.read(name_offset, linux_constants.KSYM_NAME_LEN)
+
+        idx = name_bytes.find(b"\x00")
+        if idx != -1:
+            name_bytes = name_bytes[:idx]
+
+        return name_bytes.decode("utf-8", errors="ignore")
+
+    @property
+    def value(self) -> int:
+        if self.has_member("value_offset"):
+            # kernel >= 4.19 and CONFIG_HAVE_ARCH_PREL32_RELOCATIONS=y
+            # See 7290d58095712a89f845e1bca05334796dd49ed2
+            return self._offset_to_ptr(self.value_offset)
+        elif self.has_member("value"):
+            # kernel < 4.19 or CONFIG_HAVE_ARCH_PREL32_RELOCATIONS=n
+            return self.member("value")
+
+        raise AttributeError("Unsupported kernel_symbol type implementation")
+
+    @property
+    def namespace(self) -> str:
```

Comment:
And with namespace.  This almost looks like `name` and `namespace` are  just unicode strings, I feel like these could all be handled with a `thing.name.cast("string", thing.name_offset, linux_constants.KSYM_NAME_LEN)`?

Really prefer using built in functionality than defining the same manual byte manipulation twice over...  5:S